File: packages/core/src/schemas/metadata/MetadataFunction.ts
import { IMetadataSchema } from "@typia/interface";

import { ClassProperties } from "../../typings/ClassProperties";
import { IMetadataDictionary } from "./IMetadataDictionary";
import { MetadataParameter } from "./MetadataParameter";
import { MetadataSchema } from "./MetadataSchema";

export class MetadataFunction {
  public parameters: MetadataParameter[];
  public output: MetadataSchema;
  public async: boolean;

  /** @ignore */
  private constructor(props: ClassProperties<MetadataFunction>) {
    this.parameters = props.parameters;
    this.output = props.output;
    this.async = props.async;
  }

  /** @internal */
  public static create(
    props: ClassProperties<MetadataFunction>,
  ): MetadataFunction {
    return new MetadataFunction(props);
  }

  public static from(
    json: IMetadataSchema.IFunction,
    dict: IMetadataDictionary,
  ): MetadataFunction {
    return MetadataFunction.create({
      parameters: json.parameters.map((p) => MetadataParameter.from(p, dict)),
      output: MetadataSchema.from(json.output, dict),
      async: json.async,
    });
  }

  public toJSON(): IMetadataSchema.IFunction {
    return {
      parameters: this.parameters.map((p) => p.toJSON()),
      output: this.output.toJSON(),
      async: this.async,
    };
  }
}

File: packages/core/src/schemas/metadata/MetadataFunction.go
package metadata

type MetadataFunction struct {
	Parameters []*MetadataParameter
	Output     *MetadataSchema
	Async      bool
}

type MetadataFunctionProps struct {
	Parameters []*MetadataParameter
	Output     *MetadataSchema
	Async      bool
}

func CreateMetadataFunction(props MetadataFunctionProps) *MetadataFunction {
	return &MetadataFunction{
		Parameters: append([]*MetadataParameter(nil), props.Parameters...),
		Output:     props.Output,
		Async:      props.Async,
	}
}

func MetadataFunctionFrom(json MetadataFunctionJSON, dict *IMetadataDictionary) (*MetadataFunction, error) {
	parameters := make([]*MetadataParameter, 0, len(json.Parameters))
	for _, param := range json.Parameters {
		next, err := MetadataParameterFrom(param, dict)
		if err != nil {
			return nil, err
		}
		parameters = append(parameters, next)
	}
	output, err := MetadataSchemaFrom(json.Output, dict)
	if err != nil {
		return nil, err
	}
	return CreateMetadataFunction(MetadataFunctionProps{
		Parameters: parameters,
		Output:     output,
		Async:      json.Async,
	}), nil
}

func (m *MetadataFunction) ToJSON() MetadataFunctionJSON {
	parameters := make([]MetadataParameterJSON, 0, len(m.Parameters))
	for _, param := range m.Parameters {
		parameters = append(parameters, param.ToJSON())
	}
	return MetadataFunctionJSON{
		Parameters: parameters,
		Output:     m.Output.ToJSON(),
		Async:      m.Async,
	}
}
